content/blog/real-time-smart-city-traffic-monitoring-using-microservices-based-streaming-architecture-part-2.md

```diff
@@ -0,0 +1,331 @@
+```
+
+A reading has a sensor ID, a speed in meters per second, and a time delta from time 0 (the moment the simulation starts) in seconds.
+
+My producer code simply translates the readings into a list of sensor readings ordered by time, and I transform the speed into km/s and the time into a timestamp as milliseconds from epoch.
+
+![](/img/picture4.png)
+
+Sending to the code onto the stream can be done one line at a time using standard producer code. The [code in the sample Java producer](https://docs.datafabric.hpe.com/61/MapR_Streams/code_for_the_sample_java_producer.html) works just fine.
+
+Another exciting new possibility is to use the brand new <a target='\_blank'  href='http://docs.confluent.io/2.0.0/kafka-rest/docs/index.html'>Kafka Rest Proxy</a>, which is also available on MapR from MEP 2.0 (MapR Ecosystem Pack). This means sensors can connect directly to Kafka from any language since HTTP-based REST API is a global standard.
+
+**Using the Workbench**
+
+We can login to the Workbench and login with an admin user (a user with the role “admin”) created with the `add-user.sh` script from Wildfly.
+
+Using the Workbench is beyond the scope of the article, but the general idea is to create an Organizational Unit and a Repository, and then create a project.
+
+![](/img/picture5.png)
+
+**Data Objects**
+
+We will need to create facts for the rule engine to work with. The best practice for Drools is to use a separate Maven project for your data model. For the sake of simplicity, I created them right from the workbench.
+
+![](/img/picture6.png)
+
+The Measurement is a super generic bean that models the sensor speed measurements. For the prototype, I kept it as simple as the raw data with a timestamp, a sensor ID, and a speed.
+
+![](/img/picture7.png)
+
+The Sensor bean models the sensor itself and will have an ID and the average speed of all cars that it measures over a time window defined by our rules. This average speed will be used to trigger an alert for heavy traffic. The traffic String is to mark the current traffic level which can be “NONE”, “LIGHT” or “HEAVY.”
+
+![](/img/picture8.png)
+
+**The Traffic Monitoring Rules**
+
+* create sensors for new id
+* detect heavy traffic
+* detect light traffic
+* detect normal traffic
+* get average speed at the sensor
+
+The create sensors rule ensures that there are sensor objects available in memory. This is the fact we use to know the average speed at a certain sensor.
+
+The detect heavy traffic is the key rule we want to use to send an alarm to the police if heavy traffic is detected at a certain sensor.
+
+![](/img/picture9.png)
+
+So when the average speed reaches 20 km/h or less, and the sensor isn’t already in the HEAVY traffic level, set the level to HEAVY and send an alert.
+
+Which means we need to know the average speed. Here is the rule to compute it using the Drools rule DSL (domain-specific language):
```

Comment:
Change "Which" to "This" here
![image](https://user-images.githubusercontent.com/77496261/137516631-ed4661d2-41c4-4071-a69a-4d1cafaff413.png)

Reply:
done